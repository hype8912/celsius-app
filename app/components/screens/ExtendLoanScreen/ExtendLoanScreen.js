import React, { Component } from "react";
import { TextInput, View } from "react-native";
// import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import BigNumber from "bignumber.js";
import moment from "moment";

import * as appActions from "../../../redux/actions";
import ExtendLoanScreenStyle from "./ExtendLoanScreen.styles";
import CelText from "../../atoms/CelText/CelText";
import RegularLayout from "../../layouts/RegularLayout/RegularLayout";
import CelButton from "../../atoms/CelButton/CelButton";
import formatter from "../../../utils/formatter";
import { COLOR_KEYS } from "../../../constants/COLORS";
import PredefinedAmounts from "../../organisms/PredefinedAmounts/PredefinedAmounts";
import CircleButton from "../../atoms/CircleButton/CircleButton";
import Separator from "../../atoms/Separator/Separator";
import Card from "../../atoms/Card/Card";
import { KEYBOARD_TYPE } from "../../../constants/UI";
import { SCREENS } from "../../../constants/SCREENS";
import LoadingScreen from "../LoadingScreen/LoadingScreen";
import { EXTEND_LOAN, LOAN_PAYMENT_TYPES } from "../../../constants/DATA";
import { getColor, getFontSize } from "../../../utils/styles-util";

let timeout;

export class ExtendLoanScreen extends Component {
  static propTypes = {};
  static defaultProps = {};

  inputRef = React.createRef();

  static navigationOptions = () => ({
    title: "Extend Loan",
    right: "profile",
    customCenterComponent: {steps: 3, currentStep: 1, flowProgress: true}
  });

  constructor(props) {
    super(props);
    props.actions.updateFormField("term_of_loan", "6");


    const loan = this.getLoan(props)
    this.state = {
      activePeriod: { label: `6 months`, value: 6 },
      months: 6,
      loan
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const id = this.props.navigation.getParam("id")
    if ( id && id !== prevState.loan.id) {
        this.getLoan(this.props)
    }
  }

  componentDidMount = async () => {
    const { actions } = this.props;
    actions.updateFormField("term_of_loan", 6)
    await actions.getLinkedBankAccount()
  }

  getLoan = (props) => {
    const { allLoans, navigation } = props;
    const loanId = navigation.getParam("id");
    const loan = allLoans.find(l => l.id === loanId);
    return loan
  }

  onClickPredefinedAmount = ({ value, label }) => {
    if (value === "6") {
      this.setState({
        months: EXTEND_LOAN.MIN_VALUE,
        activePeriod: {
          label
        }
      })
    }
    if (value === "36") {
      this.setState({
        months: EXTEND_LOAN.MAX_VALUE,
        activePeriod: {
          label
        },
      })
    }
  };

  handleConfirmation = (loan, bankAccountInfo, newTotal) => {
    const { actions } = this.props;
    const { months } = this.state;
    actions.updateFormFields({
      collateralCoin: loan.coin,
      ltv: loan.ltv,
      interest: loan.interest,
      loanAmount: loan.loan_amount,
      termOfLoan: `${months}`,
      bankInfo: bankAccountInfo.id ? bankAccountInfo.id : undefined,
      coin: loan.coin_loan_asset,
      loanType: loan.coin_loan_asset !== "USD" ? "STABLE_COIN_LOAN" : "USD_LOAN",
      loanId: loan.id
    });
    actions.navigateTo(SCREENS.CONFIRM_EXTEND_LOAN, { newTotal })
  }

  handleAmountChange = (newValue) => {
    const { actions } = this.props;
    if (timeout) clearTimeout(timeout);
    if (Number(newValue) < 6) {
      // Math.floor(newValue / 10)
      timeout = setTimeout(() => {
        actions.showMessage(
          "warning",
          `6 months is currently the minimum extendable period. Please adjust your extend period to proceed.`
        );
      }, 3000);
    }

    if (Number(newValue) > 36) {
      timeout = setTimeout(() => {
        actions.showMessage(
          "warning",
          `36 months is currently the maximum extendable period. Please adjust your extend period to proceed.`
        );
      }, 3000);
    }
    this.setState({ months: Number(newValue) });
  };

  extendLoanDecrement = (current) => {
    if(current === 6) return 6
    return current - 1
  };

  onClickDecrement = () => {
    const { months } = this.state;
    const data = this.extendLoanDecrement(months)
    this.setState({
      months: data
    })
  }

  extendLoanIncrement = (current) => {
    if (current === 36 ) return 36
    return current + 1
  }

  onClickIncrement = () => {
    const { months } = this.state;
    const data = this.extendLoanIncrement(months)
    this.setState({
      months: data
    })
  };

  extendLoanCalculateInterest = (loan, originatingDate, numberOfMonths) => {
    const maturityDate = originatingDate.clone().add(numberOfMonths, "month");
    const loanTermInDays = maturityDate.diff(originatingDate, "days");
    const additionalInterest = new BigNumber(loan.interest)
      .dividedBy(365)
      .multipliedBy(loanTermInDays)
      .multipliedBy(loan.loan_amount)
      .toNumber();
    const monthlyInterest = additionalInterest / numberOfMonths;
    // should be separately covered
    const amountLeft = loan.amortization_table.filter(l => l.status === "DUE" && l.type === LOAN_PAYMENT_TYPES.MONTHLY_INTEREST ).reduce((a,b) => {
      return Number(a) + Number(b.amountToPay)
    }, 0)
    const totalNewInterest = additionalInterest + amountLeft;
    return {
      totalNewInterest,
      monthlyInterest,
      additionalInterest
    }
  }

  onExtendLoanCalculateInterest = () => {
    const { months, loan } = this.state;
    const originatingDate = moment();
    const interest = this.extendLoanCalculateInterest(loan, originatingDate, months);
    return interest
  }

  render() {
    const { bankAccountInfo } = this.props;
    const { activePeriod, months, loan } = this.state;
    if (!bankAccountInfo)
      return <LoadingScreen />;
    const style = ExtendLoanScreenStyle()

    let disabled = false;
    let color = "black"

    const predefinedAmount = [
      { label: `6 months`, value: 6 },
      { label: `36 months`, value: 36 },
    ];
    const interest = this.onExtendLoanCalculateInterest()
    if (months < 6 || months > 36) {
      disabled = true
      color = getColor(COLOR_KEYS.NEGATIVE_STATE)
    }

    return (
      <RegularLayout>
        <View style={style.wrapper}>
          <View>
            <CelText align={"center"} weight={"300"}>
              How long would you like to extend loan
            </CelText>
            <PredefinedAmounts
              data={predefinedAmount}
              onSelect={this.onClickPredefinedAmount}
              activePeriod={activePeriod}
            />
          </View>

          <View style={style.circleWrapper}>
              <CircleButton
                icon={"Minus2"}
                size={30}
                onPress={() => this.onClickDecrement()}
              />
              <View style={style.input}>
                <TextInput
                  keyboardType={KEYBOARD_TYPE.NUMBER_PAD}
                  autoFocus
                  onChangeText={this.handleAmountChange}
                  maxLenght={2}
                  field={"term_of_loan"}
                  style={[style.textInput, { fontSize: getFontSize("H1"), color }]}
                >
                  {months}
                </TextInput>
                <CelText margin={"0 20 0 20"} weight={"200"} type={"H3"}>
                  {"months"}
                </CelText>
              </View>
            <CircleButton
              icon={"Plus"}
              size={30}
              onPress={() => this.onClickIncrement()}
            />
          </View>
          <View>
            <Card size={"twoThirds"} color={COLOR_KEYS.BACKGROUND}>
              <CelText type={"H6"}>{`Additional interest: ${formatter.crypto(interest.additionalInterest, loan.coin_loan_asset, { precision: 2} )}`}</CelText>
              <Separator margin={"10 0 10 0"}/>
              <CelText type={"H6"}>{`New monthly interest: ${formatter.crypto(interest.monthlyInterest, loan.coin_loan_asset, { precision: 2} )}`}</CelText>
              <Separator margin={"10 0 10 0"}/>
              <CelText type={"H6"}>{`Total interest : ${formatter.crypto(interest.totalNewInterest, loan.coin_loan_asset, { precision: 2} )}`}</CelText>
            </Card>

          </View>

          <View>
            <CelButton
              onPress={() => this.handleConfirmation(loan, bankAccountInfo, interest.totalNewInterest)}
              margin={"30 0 0 0"}
              iconRight={"IconArrowRight"}
              iconRightWidth={"20"}
              disabled={disabled}
            >
              Confirm
            </CelButton>
          </View>
        </View>

      </RegularLayout>
    );
  }
}

export default connect(
  state => ({
    currencyRates: state.currencies.currencyRatesShort,
    allLoans: state.loans.allLoans,
    bankAccountInfo: state.user.bankAccountInfo,
  }),
  dispatch => ({ actions: bindActionCreators(appActions, dispatch) })
)(ExtendLoanScreen);
