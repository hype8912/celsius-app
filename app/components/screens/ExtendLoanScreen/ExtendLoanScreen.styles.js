// import STYLES from '../../../constants/STYLES';
import { getThemedStyle } from "../../../utils/styles-util";

const base = {
  container: {
    flex: 1,
  },
  textInput: {textAlign: "center"},
  wrapper: { justifyContent: "space-around", alignItems: "center" },
  circleWrapper: {marginVertical: 30, flexDirection: "row", justifyContent: "center", alignItems: "center" },
  input: {alignItems: "center"},
};

const themed = {
  light: {},

  dark: {},

  celsius: {},
};

const ExtendLoanScreenStyle = () => getThemedStyle(base, themed);

export default ExtendLoanScreenStyle;
