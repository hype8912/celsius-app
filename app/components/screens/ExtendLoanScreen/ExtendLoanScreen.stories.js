import React from "react";
import _ from "lodash";

import { ExtendLoanScreen } from "./ExtendLoanScreen";
import ScreenStoryWrapper from "../../../../storybook/stories/ScreenStoryWrapper/ScreenStoryWrapper";
import mockUserStore from "../../../../celsius-app-creds/mock-data/mockUserStore";
import mockCurrenciesStore from "../../../../celsius-app-creds/mock-data/mockCurrenciesStore";
import mockLoansStore from "../../../../celsius-app-creds/mock-data/mockLoansStore";

const initialState = {
  user: {
    profile: mockUserStore.profile.postman13,
    appSettings: mockUserStore.appSettings.postman13,
    bankAccountInfo: mockUserStore.bankAccountInfo.postman13
  },
  currencies: {
    rates: mockCurrenciesStore.rates,
    currencyRatesShort: mockCurrenciesStore.currencyRatesShort,
  },
  loans: {
    allLoans: mockLoansStore.allLoans.enimalnowlt3,
  },
  forms: {
    formData: {
      collateralCoin: "BTC",
      ltv: "0.33",
      interest:  "0.0395",
      loanAmount:"1500",
      termOfLoan: 6,
      bankInfo: "7bcd5ccb-0f01-4676-9fec-286b1637288a",
      coin: "USDC",
      loanType:"STABLE_COIN_LOAN" ,
      loanId: 1096
    },
  },
};

const navigationProps = {
  id: initialState.loans.allLoans[0].id,
};

const regular = () => {
  const state = _.cloneDeep(initialState);

  return (
    <ScreenStoryWrapper
      screenName="ExtendLoanScreen"
      screen={ ExtendLoanScreen }
      state={ state }
      navigationProps={navigationProps}
    />
  );
};

export default {
  regular,
};
