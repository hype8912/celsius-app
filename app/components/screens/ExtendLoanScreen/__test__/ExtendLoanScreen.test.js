import React from "react"
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from "enzyme"
// import moment from "moment";
//
// import store from '../../../../redux/store';
// import formatter from "../../../../utils/formatter";
import ExtendLoanScreen from "../ExtendLoanScreen"
import mockLoansStore from "../../../../../celsius-app-creds/mock-data/mockLoansStore";
import mockUserStore from "../../../../../celsius-app-creds/mock-data/mockUserStore";
import mockCurrenciesStore from "../../../../../celsius-app-creds/mock-data/mockCurrenciesStore";

// mount

configure({adapter: new Adapter()});

const Context = React.createContext()

const baseProps = {
  allLoans: mockLoansStore.allLoans.enimalnowlt3,
  bankAccountInfo: mockUserStore.bankAccountInfo.postman13,
  currencyRatesShort: mockCurrenciesStore.currencyRatesShort,
}

const wrapper = shallow(
  <Context.Provider store={baseProps}>
    <ExtendLoanScreen/>
  </Context.Provider>
);
// instance()
describe(`extendLoanCalculateInterest`, () => {
  it(`should increment number by one`, () => {
    // console.log("wrapper", Object.keys(wrapper))
    // const result = wrapper.extendLoanIncrement(1)
    // expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();
    // expect(result).toEqual(2)
    //     // const result = wrapper.extendLoanIncrement(1)
    //     expect(wrapper.find().extendLoanIncrement(1)).toEqual(2)
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();
  })
})

describe(`extendLoanCalculateInterest`, () => {
  it(`should decrement number by one`, () => {

    // const result = wrapper.extendLoanDecrement(2)
    // expect(result).toEqual(1)
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

describe(`extendLoanCalculateInterest`, () => {
  it(`if value === 36 should return 36`, () => {

    // const result = wrapper.extendLoanIncrement(36)
    // expect(result).toEqual(36)
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

// if undefined should show error

describe(`extendLoanCalculateInterest`, () => {
  it(`if value === 6 should return 6`, () => {

    // const result = wrapper.extendLoanDecrement(6)
    // expect(result).toEqual(6)
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

describe(`extendLoanCalculateInterest`, () => {
  it(`Should Calculate Additional Interest in Crypto`, () => {

    // const result = wrapper.calculateAdditionalInterest(50, 2, "BTC")
    // expect(result).toEqual(formatter.crypto(25, "BTC", {precisions: 2}))
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

describe(`extendLoanCalculateInterest`, () => {
  it(`Should Calculate Additional Interest in USD`, () => {

    // const result = wrapper.calculateAdditionalInterest(50, 1, "USD")
    // expect(result).toEqual(formatter.crypto(50, "USD", {precisions: 2}))
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

// broj meseci i broj dana u razlicitom opsegu varira
describe(`extendLoanCalculateInterest`, () => {
  it(`should return total new interest, monthly interest, additional interest`, () => {

    // const result = wrapper.extendLoanCalculateInterest({interest: `0.01`, loan_amount: `1000`}, moment(), 5)
    // expect(result).toEqual(1)
    expect(wrapper.find(ExtendLoanScreen)).toMatchSnapshot();

  })
})

// undefined



// // handleAmountChange and press predefined value
// describe(`extendLoanCalculateInterest`, () => {
//   it(`Should forward any number it receives`, () => {
//
//     const result = wrapper.handleAmountChange(36)
//     expect(result).toEqual(36)
//   })
// })
//
// // getLoan? get loans from mock and send random id from that object
// describe(`extendLoanCalculateInterest`, () => {
//   it(`Should return loan`, () => {
//
//     const result = wrapper.getLoan()
//     expect(result).toEqual(6)
//   })
// })


// import mockLoansStore from "../../../../../celsius-app-creds/mock-data/mockLoansStore";
// import mockUserStore from "../../../../../celsius-app-creds/mock-data/mockUserStore";
// import mockCurrenciesStore from "../../../../../celsius-app-creds/mock-data/mockCurrenciesStore";
//
// configure({adapter: new Adapter()});
//

//
