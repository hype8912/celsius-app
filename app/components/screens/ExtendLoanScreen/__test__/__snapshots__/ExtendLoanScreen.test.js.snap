// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extendLoanCalculateInterest Should Calculate Additional Interest in Crypto 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest Should Calculate Additional Interest in USD 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest if value === 6 should return 6 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest if value === 36 should return 36 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest should decrement number by one 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest should increment number by one 1`] = `<Connect(ExtendLoanScreen) />`;

exports[`extendLoanCalculateInterest should return total new interest, monthly interest, additional interest 1`] = `<Connect(ExtendLoanScreen) />`;
